### ISSUE TITLE
Incorporate Google's CityHash successor: FarmHash

### ISSUE THREAD

**Issue by NickStrupat** [brandondahler/Data.HashFunction#32]
@nickbabcock has a .NET implementation at

https://github.com/nickbabcock/Farmhash.Sharp

**Comment by nickbabcock** [comment]
Thanks for the shout out. I won't speak for brandon, but I believe our two libraries may have a different focus. Farmhash.Sharp emphasizes speed whereas Data.HashFunction emphasizes completeness, safety, and ergonomics. This is apparent in benchmarks where Farmhash.Sharp exhibits 10x throughput over 64bit Cityhash from Data.HashFunction*. I've included a heatmap of relative throughput with Farmhash.Sharp and Data.HashFunction highlighted.

While I certainly encourage Data.HashFunction to incorporate an implementation of Farmhash, if your sole purpose for a non-cryptographic hash function is throughput, I'm skeptical a safe implementation in Data.HashFunction will match performance.

![image](https://user-images.githubusercontent.com/2106129/33146936-b1ea9552-cf8b-11e7-93ee-90a252140f17.png)

*For 64bit RYU JIT platform -- arguably the most common platform

**Comment by NickStrupat** [comment]
Thanks @nickbabcock, great info here. As an aside, I wonder how much of that 10x throughput advantage of FarmHash over CityHash comes from differences in the underlying algorithm.

Back on topic though, after taking a peek into this libraries' code I see totally managed implementation. So I agree with you on the different approaches. Alas, if either of us come across a managed FarmHash implementation let's meet back here, deal?

**Comment by nickbabcock** [comment]
Deal 😄 

Farmhash.Sharp shouldn't be 10x faster than an equivalent .NET Cityhash implementation as both are very similar. Farmhash (by google) was designed to take advantage of modern hardware (SSE 4.2 instructions). I'm banking on .NET to eventually implement these intrinsics -- then you'll see Farmhash.Sharp throughput dramatically increase for large payloads.

**Comment by brandondahler** [comment]
@nickbabcock is exactly right about the differences between the overall focus between the two libraries.  

I'd love to add an implementation of FarmHash to Data.HashFunction, but in general if one is looking for the absolute fastest implementation of a given hash function, they will likely be able to find a more specialized library that could provide better performance.

Keeping this ticket open to track progress on said implementation.

**Comment by brandondahler** [comment]
@nickbabcock @NickStrupat I was able to implement the Fingerprint32, Fingerprint64, and Fingerprint128 methods of FarmHash.  


@nickbabcock I went to create a binding implementation for FarmHash.Sharp but found that you're not strong name signing your assemblies -- I don't want to start the signing versus non-signing debate that's been done many times.  I personally have chosen to sign Data.HashFunction with a project-specific private key that has been checked in publicly (and therefore is no longer private).  

I can create my binding implementation as non-signed, but you might consider signing your assemblies, even if it is just with a unprotected signing key. See https://docs.microsoft.com/en-us/dotnet/framework/app-domains/strong-named-assemblies , for benefits and their OSS recommendation (literally the last sentence).
